storagemarket/smtestutil/mocks.go

```diff
@@ -94,27 +99,31 @@ func (mb *MinerStubBuilder) SetupAllBlocking() *MinerStubBuilder {
 }
 
 func (mb *MinerStubBuilder) SetupPublish(blocking bool) *MinerStubBuilder {
-	publishCid := testutil.GenerateCid()
-
 	mb.stub.lk.Lock()
 	if blocking {
 		mb.stub.unblockPublish[mb.dp.DealUUID] = make(chan struct{})
 	}
 	mb.stub.lk.Unlock()
 
-	mb.stub.MockDealPublisher.EXPECT().Publish(gomock.Any(), gomock.Eq(mb.dp.DealUUID), gomock.Eq(mb.dp.ClientDealProposal)).DoAndReturn(func(_ context.Context, _ uuid.UUID, _ market2.ClientDealProposal) (cid.Cid, error) {
+	mb.stub.MockDealPublisher.EXPECT().Publish(gomock.Any(), gomock.Eq(mb.dp.DealUUID), gomock.Eq(mb.dp.ClientDealProposal)).DoAndReturn(func(ctx context.Context, _ uuid.UUID, _ market2.ClientDealProposal) (cid.Cid, error) {
 		mb.stub.lk.Lock()
 		ch := mb.stub.unblockPublish[mb.dp.DealUUID]
 		mb.stub.lk.Unlock()
 		if ch != nil {
-			<-ch
+			select {
+			case <-ctx.Done():
+				return cid.Undef, ctx.Err()
+			case <-ch:
+			}
+
+		}
+		if ctx.Err() != nil {
+			return cid.Undef, ctx.Err()
```

Comment:
I don't think you need to check `if ctx.Err() != nil` because you're already selecting against `<-ctx.Done()` which does the same thing.

Reply:
Yeah but that select is guarded by an if.